datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxClusterUpgradeService.java

```diff
@@ -0,0 +1,74 @@
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.datalake.controller.exception.BadRequestException;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.flow.SdxReactorFlowManager;
+import com.sequenceiq.datalake.service.sdx.SdxService;
+import com.sequenceiq.flow.api.model.FlowIdentifier;
+
+@Component
+public class SdxClusterUpgradeService {
+
+    private static final long WORKSPACE_ID = 0L;
+
+    @Inject
+    private StackV4Endpoint stackV4Endpoint;
+
+    @Inject
+    private SdxService sdxService;
+
+    @Inject
+    private SdxReactorFlowManager sdxReactorFlowManager;
+
+    public UpgradeOptionsV4Response checkForClusterUpgradeByName(String name) {
+        return stackV4Endpoint.checkForClusterUpgradeByName(0L, name);
+    }
+
+    public UpgradeOptionsV4Response checkForStackUpgradeByCrn(String userCrn, String crn) {
+        SdxCluster sdxCluster = sdxService.getByCrn(userCrn, crn);
+        return stackV4Endpoint.checkForClusterUpgradeByName(WORKSPACE_ID, sdxCluster.getClusterName());
+    }
+
+    public FlowIdentifier triggerUpgradeByName(String userCrn, String clusterName, String imageId) {
+        SdxCluster cluster = sdxService.getSdxByNameInAccount(userCrn, clusterName);
+        validateImageId(clusterName, imageId);
+        MDCBuilder.buildMdcContext(cluster);
+        return sdxReactorFlowManager.triggerDatalakeClusterUpgradeFlow(cluster.getId(), imageId);
+    }
+
+    public FlowIdentifier triggerUpgradeByCrn(String userCrn, String clusterCrn, String imageId) {
+        SdxCluster cluster = sdxService.getByCrn(userCrn, clusterCrn);
+        validateImageId(cluster.getClusterName(), imageId);
+        MDCBuilder.buildMdcContext(cluster);
+        return sdxReactorFlowManager.triggerDatalakeClusterUpgradeFlow(cluster.getId(), imageId);
+    }
+
+    private void validateImageId(String clusterName, String imageId) {
+        UpgradeOptionsV4Response upgradeOptionsV4Response = checkForClusterUpgradeByName(clusterName);
+        List<ImageInfoV4Response> upgradeCandidates = upgradeOptionsV4Response.getUpgradeCandidates();
+        if (Objects.isNull(upgradeCandidates) || upgradeCandidates.isEmpty()) {
+            throw new BadRequestException(String.format("There is no compatible image to upgrade for stack %s", clusterName));
+        } else if (upgradeCandidates.stream().noneMatch(imageInfoV4Response -> imageInfoV4Response.getImageId().equalsIgnoreCase(imageId))) {
+            String candidates = upgradeCandidates.stream().map(ImageInfoV4Response::getImageId).collect(Collectors.joining(","));
+            throw new BadRequestException(String.format("The given image (%s) is not eligible for upgrading the cluster. "
+                    + "Please choose an id from the following image(s): %s", imageId, candidates));
+        } else if (Strings.isNotEmpty(upgradeOptionsV4Response.getReason())) {
```

Comment:
Use the apache commons `StringUtils` please. `Strings` is from a logging library.

Reply:
Ok.